repos/system_upgrade/el7toel8/actors/commonleappdracutmodules/files/dracut/85sys-upgrade-redhat/do-upgrade.sh

```diff
@@ -17,6 +17,149 @@ NSPAWN_OPTS="--capability=all --bind=/sys --bind=/dev --bind=/dev/pts --bind=/ru
+
+want_inband_dump() {
+    #
+    # True if dump collection is needed given leapp exit status $1 and kernel option
+    #
+    local leapp_es=$1
+    local mode
+    local kopt
+    kopt=$(getarg 'rd.upgrade.inband')
+    case $kopt in
+        always|never|onerror)   mode="$kopt" ;;
+        "")                     mode="never" ;;
+        *)  warn "ignoring unknown value of rd.upgrade.inband (dump will be disabled): '$kopt'"
+            return 2 ;;
+    esac
+    case $mode:$leapp_es in
+        always:*)   return 0 ;;
+        never:*)    return 1 ;;
+        onerror:0)  return 1 ;;
+        onerror:*)  return 0 ;;
+    esac
+}
+
+ibdmp() {
+    #
+    # Dump tarball $1 in base64 to stdout
+    #
+    # Tarball is encoded in a way that:
+    #
+    #   * final data can be printed to plain text terminal,
+    #   * tarball can be restored by scanning the saved
+    #     terminal output,
+    #   * corruptions caused by extra terminal noise
+    #     (extra lines, extra characters within lines,
+    #     line splits..) can be corrected.
+    #
+    # That is,
+    #
+    #   1. encode tarball using base64
+    #
+    #   2. pre-pend line `chunks=CHUNKS,md5=MD5` where
+    #      MD5 is the MD5 digest of original tarball and
+    #      CHUNKS is number of upcoming Base64 chunks
+    #
+    #   3. decorate each chunk with prefix `N:` where
+    #      N is number of given chunk.
+    #
+    #   4. Finally print all lines (pre-pended "header"
+    #      line and all chunks) several times, where
+    #      every iteration should be prefixed by
+    #      `_ibdmp:I/TTL|` and suffixed by `|`.
+    #      where `I` is iteration number and `TTL` is
+    #      total iteration numbers.
+    #
+    # Decoder should look for strings like this:
+    #
+    #     _ibdmp:I/J|CN:PAYLOAD|
+    #
+    # where I, J and CN are integers and PAYLOAD is a slice of a
+    # base64 string.
+    #
+    # Here, I represents number of iteration, J total of iterations
+    # ($IBDMP_ITER), and CN is number of given chunk within this
+    # iteration.  CN goes from 1 up to number of chunks (CHUNKS)
+    # predicted by header.
+    #
+    # Each set corresponds to one dump of the tarball and error
+    # correction is achieved by merging sets using these rules:
+    #
+    #    1. each set has to contain header (`chunks=CHUNKS,
+    #       md5=MD5`) prevalent header wins.
+    #
+    #    2. each set has to contain number of chunks
+    #       as per header
+    #
+    #    3. chunks are numbered so they can be compared across
+    #       sets; prevalent chunk wins.
+    #
+    # Finally the merged set of chunks is decoded as base64.
+    # Resulting data has to match md5 sum or we're hosed.
+    #
+    local tarball=$1
+    local tmp=$LEAPP_DEBUG_TMP/ibdmp
+    local md5
+    local i
+    mkdir -p "$tmp"
+    base64 -w "$IBDMB_CHUNKSIZE" "$tarball" > "$tmp/b64"
+    md5=$(md5sum "$tarball" | sed 's/ .*//')
+    chunks=$(wc -l <"$tmp/b64")
+    (
+        set +x
```

Comment:
This doesn't need to be here, right?

Reply:
I think it does.  --  Note that `+x` *disables* the shell debug mode.

So if I drop it, the shell will re-print every upcoming arithmetic and `echo`, making the log **massive**.

Reply:
oh right, the perks of shell using `+` to disable things and `-` to disable things :D

Reply:
Yeah, no blame though, POSIX might be older than modern use of `+` and `-` in mathematics :D